### ISSUE TITLE
Parallel Bundle Install extension overwrites "typeset" built-in and breaks compinit (on some systems/versions of ZSH)

### ISSUE THREAD

**Issue by dmotles** [zsh-users/antigen#568]
Hi Antigen Devs,

I spent a few hours banging my head against this problem, and in the spirit of giving you the most information possible, forgive me for the text wall. But I feel that you will want to fix this so other users don't start complaining :)

I git cloned the repo on the develop branch (since its the default branch). I added the following config to my .zshrc.

```
source ~/antigen/antigen.zsh
antigen use oh-my-zsh
antigen bundle zsh-users/zsh-syntax-highlighting
antigen theme bureau
antigen apply
```

where ~/antigen is where I am cloning this repo from.

This worked fine on my mac (zsh 5.2 (x86_64-apple-darwin16.0) macOS Sierra 10.12.5)

It does *not* work on my ubuntu VM running Ubuntu Trusty 14.04 (zsh 5.0.2 (x86_64-pc-linux-gnu))

When ZSH starts up, first error:

```
-antigen-hook-antigen-apply:zcompile:11: can't open file: /home/dmotles/.antigen/.zcompdump
```

I noticed that the system `/etc/zsh/zshrc` applies compinit... 

```
cat /etc/zsh/zshrc 
# /etc/zsh/zshrc: system-wide .zshrc file for zsh(1).
#
# This file is sourced only for interactive shells. It
# should contain commands to set up aliases, functions,
# options, key bindings, etc.
#
# Global Order: zshenv, zprofile, zshrc, zlogin

READNULLCMD=${PAGER:-/usr/bin/pager}

# An array to note missing features to ease diagnosis in case of problems.
typeset -ga debian_missing_features

if [[ -z "$DEBIAN_PREVENT_KEYBOARD_CHANGES" ]] &&
   [[ "$TERM" != 'emacs' ]]
then

    typeset -A key
    key=(
        BackSpace  "${terminfo[kbs]}"
        Home       "${terminfo[khome]}"
        End        "${terminfo[kend]}"
        Insert     "${terminfo[kich1]}"
        Delete     "${terminfo[kdch1]}"
        Up         "${terminfo[kcuu1]}"
        Down       "${terminfo[kcud1]}"
        Left       "${terminfo[kcub1]}"
        Right      "${terminfo[kcuf1]}"
        PageUp     "${terminfo[kpp]}"
        PageDown   "${terminfo[knp]}"
    )

    function bind2maps () {
        local i sequence widget
        local -a maps

        while [[ "$1" != "--" ]]; do
            maps+=( "$1" )
            shift
        done
        shift

        sequence="${key[$1]}"
        widget="$2"

        [[ -z "$sequence" ]] && return 1

        for i in "${maps[@]}"; do
            bindkey -M "$i" "$sequence" "$widget"
        done
    }

    bind2maps emacs             -- BackSpace   backward-delete-char
    bind2maps       viins       -- BackSpace   vi-backward-delete-char
    bind2maps             vicmd -- BackSpace   vi-backward-char
    bind2maps emacs             -- Home        beginning-of-line
    bind2maps       viins vicmd -- Home        vi-beginning-of-line
    bind2maps emacs             -- End         end-of-line
    bind2maps       viins vicmd -- End         vi-end-of-line
    bind2maps emacs viins       -- Insert      overwrite-mode
    bind2maps             vicmd -- Insert      vi-insert
    bind2maps emacs             -- Delete      delete-char
    bind2maps       viins vicmd -- Delete      vi-delete-char
    bind2maps emacs viins vicmd -- Up          up-line-or-history
    bind2maps emacs viins vicmd -- Down        down-line-or-history
    bind2maps emacs             -- Left        backward-char
    bind2maps       viins vicmd -- Left        vi-backward-char
    bind2maps emacs             -- Right       forward-char
    bind2maps       viins vicmd -- Right       vi-forward-char

    # Make sure the terminal is in application mode, when zle is
    # active. Only then are the values from $terminfo valid.
    if (( ${+terminfo[smkx]} )) && (( ${+terminfo[rmkx]} )); then
        function zle-line-init () {
            emulate -L zsh
            printf '%s' ${terminfo[smkx]}
        }
        function zle-line-finish () {
            emulate -L zsh
            printf '%s' ${terminfo[rmkx]}
        }
        zle -N zle-line-init
        zle -N zle-line-finish
    else
        for i in {s,r}mkx; do
            (( ${+terminfo[$i]} )) || debian_missing_features+=($i)
        done
        unset i
    fi

    unfunction bind2maps

fi # [[ -z "$DEBIAN_PREVENT_KEYBOARD_CHANGES" ]] && [[ "$TERM" != 'emacs' ]]

zstyle ':completion:*:sudo:*' command-path /usr/local/sbin \
                                           /usr/local/bin  \
                                           /usr/sbin       \
                                           /usr/bin        \
                                           /sbin           \
                                           /bin            \
                                           /usr/X11R6/bin

(( ${+aliases[run-help]} )) && unalias run-help
autoload -Uz run-help

# If you don't want compinit called here, place the line
# skip_global_compinit=1
# in your $ZDOTDIR/.zshenv or $ZDOTDIR/.zprofile
if [[ -z "$skip_global_compinit" ]]; then
  autoload -U compinit
  compinit
fi
```

I decided to disable it by setting  `skip_global_compinit=1` in my `.zshenv` to let antigen do the compinit stuff.

That yielded an interesting error next time I started ZSH:

```
compinit:493: _i_test: assignment to invalid subscript range
```

I enabled shell tracing at the top of my .zshrc (`set -x`) to figure out what was going on, and inspected the compinit function.

I noticed something weird in the traceback...

```
+compinit:485> typeset -A _i_test
+typeset:1> -antigen-interactive-mode
+-antigen-interactive-mode:1> WARN '-antigen-interactive-mode: file:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:loadautofunc:shfunc:shfunc $_ANTIGEN_INTERACTIVE = '
+WARN:1> local 'PREFIX=[WRN][1498177056.9634315968]'
+WARN:2> echo '[WRN][1498177056.9634315968] /home/dmotles/antigen/bin/antigen.zsh:246\n[WRN][1498177056.9634315968] -antigen-interactive-mode: file:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:loadautofunc:shfunc:shfunc $_ANTIGEN_INTERACTIVE = '
+-antigen-interactive-mode:2> [[ '' !=  ]]
+-antigen-interactive-mode:7> [[ file:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:loadautofunc:shfunc:shfunc == toplevel* || file:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:shfunc:loadautofunc:shfunc:shfunc == cmdarg* ]]
```

Huh? compinit called the built-in `typeset` command and the `-antigen-interactive-mode` function was called from that? I checked and lo and behold:

```
+-zsh:2> which typeset
typeset () {
	if -antigen-interactive-mode
	then
		return 1
	fi
}
```

That seems... broken. I checked against my mac, and typset is still set to the built-in.

I started adding `which typeset` all over the place, and noticed that typeset gets overridden after the initial `source ~/antigen/antigen.zsh`. I added some more `which typeset` into the `bin/antigen.zsh` on my system, and saw that it gets overridden *after* the parallel extension is loaded in `antigen-ext-init`.

```
+antigen-ext:9> _ANTIGEN_EXTENSIONS+=( lock ) 
+antigen-ext-init:6> echo '------------------\t\t\t\t\tMOTLES DEBUG>>> lock'
------------------					MOTLES DEBUG>>> lock
+antigen-ext-init:7> which typeset
typeset: shell built-in command
+antigen-ext-init:3> ext=parallel
+antigen-ext-init:5> ((  1  ))
+antigen-ext-init:5> antigen-ext parallel
+antigen-ext:1> local 'ext=parallel'
+antigen-ext:2> local 'func=-antigen-parallel-init'
+antigen-ext:3> ((  1 && 0 == 0  ))
+antigen-ext:4> eval -antigen-parallel-init
+(eval):1> -antigen-parallel-init
+-antigen-parallel-init:1> WARN 'Init parallel extension' PARALLEL
+WARN:1> local 'PREFIX=[WRN][1498177562.1505901814]'
+WARN:2> echo '[WRN][1498177562.1505901814] /home/dmotles/antigen/bin/antigen.zsh:1658\n[WRN][1498177562.1505901814] Init parallel extension' PARALLEL
+antigen-ext:5> local 'ret=0'
+antigen-ext:6> WARN '-antigen-parallel-init return code was 0'
+WARN:1> local 'PREFIX=[WRN][1498177562.1507265568]'
+WARN:2> echo '[WRN][1498177562.1507265568] /home/dmotles/antigen/bin/antigen.zsh:1554\n[WRN][1498177562.1507265568] -antigen-parallel-init return code was 0'
+antigen-ext:7> ((  0 == 0  ))
+antigen-ext:8> LOG 'LOADED EXTENSION parallel' EXT
+LOG:1> local 'PREFIX=[LOG][1498177562.1508214474]'
+LOG:2> echo '[LOG][1498177562.1508214474] /home/dmotles/antigen/bin/antigen.zsh:1556\n[LOG][1498177562.1508214474] LOADED EXTENSION parallel' EXT
+antigen-ext:9> -antigen-parallel-execute
+-antigen-parallel-execute:1> WARN 'Exec parallel extension' PARALLEL
+WARN:1> local 'PREFIX=[WRN][1498177562.1509151459]'
+WARN:2> echo '[WRN][1498177562.1509151459] /home/dmotles/antigen/bin/antigen.zsh:1666\n[WRN][1498177562.1509151459] Exec parallel extension' PARALLEL
+-antigen-parallel-execute:62> antigen-add-hook antigen-apply antigen-apply-parallel pre once
+antigen-add-hook:1> local 'target=antigen-apply' 'hook=antigen-apply-parallel' 'type=pre' 'mode=once'
+antigen-add-hook:3> ((  ! 1  ))
+antigen-add-hook:8> ((  ! 1  ))
+antigen-add-hook:13> [[ antigen-apply-lock ==  ]]
+antigen-add-hook:16> _ANTIGEN_HOOKS[$target]=antigen-apply-lock:antigen-apply-parallel 
+antigen-add-hook:19> _ANTIGEN_HOOKS_META[$hook]='target antigen-apply type pre mode once called 0' 
+antigen-add-hook:22> local 'hook_function=-antigen-hook-antigen-apply'
+antigen-add-hook:23> ((  ! 1  ))
+antigen-add-hook:34> return 0
+-antigen-parallel-execute:74> antigen-add-hook antigen-apply antigen-apply-parallel-execute replace once
+antigen-add-hook:1> local 'target=antigen-apply' 'hook=antigen-apply-parallel-execute' 'type=replace' 'mode=once'
+antigen-add-hook:3> ((  ! 1  ))
+antigen-add-hook:8> ((  ! 1  ))
+antigen-add-hook:13> [[ antigen-apply-lock:antigen-apply-parallel ==  ]]
+antigen-add-hook:16> _ANTIGEN_HOOKS[$target]=antigen-apply-lock:antigen-apply-parallel:antigen-apply-parallel-execute 
+antigen-add-hook:19> _ANTIGEN_HOOKS_META[$hook]='target antigen-apply type replace mode once called 0' 
+antigen-add-hook:22> local 'hook_function=-antigen-hook-antigen-apply'
+antigen-add-hook:23> ((  ! 1  ))
+antigen-add-hook:34> return 0
+antigen-ext:9> _ANTIGEN_EXTENSIONS+=( parallel ) 
+antigen-ext-init:6> echo '------------------\t\t\t\t\tMOTLES DEBUG>>> parallel'
------------------					MOTLES DEBUG>>> parallel
+antigen-ext-init:7> which typeset
typeset () {
	if -antigen-interactive-mode
	then
		return 1
	fi
}
```

I noticed that parallel bundle install was added recently, and started checking out older commits. I found that this problem appears to go away if you go back to 09e589f69d489dccd9ebf1c1a66831a6e3d818c9, before the parallelization change was added.

I looked all around the code but to be honest, I have no idea why this happens. There is no function explicitly defined anywhere called "typeset". My zsh-foo is not strong enough to understand this. I hope that perhaps you guys might be able to offer some insight as to why this is happening.

As for now, I will probably leave the develop branch and just try to use a stable release... :)

**Comment by desyncr** [comment]
I did face this same issue on older zsh versiones and made a hack to avoid it as I could not find the core issue.

Gonna take a look.

**Comment by desyncr** [comment]
Can confirm this happens on `5.0.x` (`5.1` and above works).

**Comment by desyncr** [comment]
@dmotles Should be fixed in v2.2.1. Thanks for reporting!
